reframe/frontend/statistics.py

```diff
@@ -1,89 +1,98 @@
+from itertools import chain
+
 import reframe.core.debug as debug
+import reframe.core.exceptions as exc
 
 
 class TestStats:
     """Stores test case statistics."""
 
-    def __init__(self, tasks=[]):
-        if not hasattr(tasks, '__iter__'):
-            raise TypeError('expected an iterable')
-
-        # Store test cases per partition internally
-        self._tasks_bypart = {}
-        for t in tasks:
-            partition = t.check.current_partition
-            partname = partition.fullname if partition else 'None'
-            tclist = self._tasks_bypart.setdefault(partname, [])
-            tclist.append(t)
+    def __init__(self):
+        self._tasks = []
+        self._try_num = -1
 
     def __repr__(self):
         return debug.repr(self)
 
-    def num_failures(self, partition=None):
-        num_fails = 0
-        if partition:
-            num_fails += len([
-                t for t in self._tasks_bypart[partition] if t.failed
-            ])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_fails += len([t for t in tclist if t.failed])
-
-        return num_fails
-
-    def num_failures_stage(self, stage):
-        num_fails = 0
-        for tclist in self._tasks_bypart.values():
-            num_fails += len([t for t in tclist if t.failed_stage == stage])
-
-        return num_fails
-
-    def num_cases(self, partition=None):
-        num_cases = 0
-        if partition:
-            num_cases += len(self._tasks_bypart[partition])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_cases += len(tclist)
-
-        return num_cases
-
-    def failure_report(self):
+    def next_try(self, try_num):
+        # TODO: RAISE DIFFERENT error?
+        if try_num != self._try_num + 1:
+            raise exc.ReframeFatalError('try_num out of sync')
+
+        self._try_num = try_num
+        self._tasks.append([])
+
+    def add_task(self, task):
+        self._tasks[self._try_num].append(task)
+
+    # The try_num can be indexed from the end (e.g. -1 for the last retry)
+    def _convert_try_num(self, try_num):
+        return self._try_num + try_num + 1 if try_num < 0 else try_num
+
+    def num_success_all_retries(self):
+        tasks_retries = list(chain.from_iterable(self._tasks[1:]))
+        return len([t for t in tasks_retries if not t.failed])
+
+    def num_success(self, try_num=0):
+        try_num = self._convert_try_num(try_num)
+        return len([t for t in self._tasks[try_num] if not t.failed])
+
+    def num_failures(self, try_num=0):
```

Comment:
The second argument of this method and all similar ones must be named as `run`. It should be by default `-1` so as to give you always the results of the last run. **BUT** you must check for index/key errors and raise an error accordingly:

```python
def _get_tasks(run):
    try:
        return self._tasks[run]
    except KeyError: # if you go for dictionaries
        raise StatisticsError('no such run: %s' % run)
```

Reply:
The word `run` is already used in different contexts with different meanings. Thus, I would rather go for something like `retry` or `try` or anything similar but not `run`. I think (re-)try makes it pretty clear what we are talking about, while `run` is so general that it could mean many things.

Reply:
I don't like the word `try`, because we are introducing a new term for sth that we already have one. The `Runner` _runs_ as a set of checks and we have a _run action_ as well to describe the same this action. So, since now we can run multiple times some tests, I see pretty much natural to count the number of _runs_ you do. The number of retries is always `num_runs - 1`. So, in summary, I would either count the _runs_ or the _retries_.

Reply:
OK, in agreement with the variable `_current_run`, I use here now `run`. Done.